cmd/better-retester/retester.go

```
@@ -3,58 +3,76 @@ package main
 	configGetter config.Getter
 
+	// changedFiles caches the names of files changed by PRs.
+	// Cache entries expire if they are not used during a sync loop.
+	changedFiles *changedFilesAgent
+
 	logger *logrus.Entry
 
 	usesGitHubApp bool
 }
 
-func newController(ghClient githubClient, cfg config.Getter, usesApp bool) *retestController {
+func newController(ghClient githubClient, cfg config.Getter, gitClient git.ClientFactory, usesApp bool) *retestController {
 	return &retestController{
-		ghClient:      ghClient,
-		configGetter:  cfg,
+		ghClient:     ghClient,
+		gitClient:    gitClient,
+		configGetter: cfg,
+		changedFiles: &changedFilesAgent{
+			ghc:             ghClient,
+			nextChangeCache: make(map[changeCacheKey][]string),
+		},
 		logger:        logrus.NewEntry(logrus.StandardLogger()),
 		usesGitHubApp: usesApp,
 	}
 }
 
 func (c *retestController) sync() error {
+	defer c.changedFiles.prune()
 	// Input: Tide Config
-	// Output: A list of PRs that would merge but have failing jobs
+	// Output: A list of PRs that are filter out by the queries in Tide's config
 	candidates, err := findCandidates(c.configGetter, c.ghClient, c.usesGitHubApp, c.logger)
 	if err != nil {
 		return err
 	}
 
-	logrus.Infof("Found %d candidates for retest", len(candidates))
-	for key, pr := range candidates {
-		logrus.Infof("Candidate PR: %s (https://github.com/%s/%s/pull/%d", key, pr.Repository.Owner.Login, pr.Repository.Name, pr.Number)
+	logrus.Infof("Found %d candidates for retest (pass label criteria, fail some tests)", len(candidates))
+	for _, pr := range candidates {
+		logrus.Infof("Candidate PR: https://github.com/%s/%s/pull/%d", pr.Repository.Owner.Login, pr.Repository.Name, pr.Number)
 	}
 
-	// Input: A list of PRs that would merge but have failing jobs
-	// Output: A subset of input PRs whose jobs are actually required for merge
-	candidates = atLeastOneFailingRequiredJob(candidates)
+	candidates, _ = c.atLeastOneRequiredJob(candidates)
```

I also need to not ignore this error